### PR TITLE
Change user updates to require explicit reactivation

### PR DESCRIPTION
**Pull request by senior**
Previously, attempting to update a user with the same email address as
an existing inactive user would trigger an implicit user reactivation
flow. This commit changes that implicit action to an explicit
one. Attempting to update an existing inactive user will now
fail. The `GET /` endpoint will not include all users, active and
inactive so that inactive users can be shown. This also creates a new
`/user/:id/reactivate` endpoint that will invoke the previously
implicit reactivation feature.

Based on what's in this PR and the designs from https://github.com/metabase/metabase/issues/4936#issuecomment-344109124 I think below is what still needs to be done.
- [x] Update the `People` page to include a Deactivated tab. Users from the `GET /api/users/` call will now include an `is_active` boolean and all users.
- [x] On the deactivated tab, add the reactivate link that uses the new `PUT /api/user/:id/reactivate` endpoint (along with the dialog)
- [x] The backend returns a 400 if the user attempts to update a user's email address to an existing address, but the frontend isn't propagating this error, that needs to be updated (see https://github.com/metabase/metabase/issues/7312)

### CODE REVIEW

**Comment by tlrobinson** [comment]
This is probably going to cause deactivated users to appear in the pulses recipient picker (and maybe other places?). Not sure if we should filter there or have a query parameter to `include_deactivated` or something.

**Comment by tlrobinson** [comment]
@senior is there a way the frontend can get when a deactivated user was deactivated? That's in the designs.

File: src/metabase/api/user.clj [hunk 1]
```diff
@@ -89,6 +83,17 @@
                                    is_superuser)))
   (User id))
 
+(api/defendpoint PUT "/:id/reactivate"
```
Comment (tlrobinson) [inline src/metabase/api/user.clj]: I think this should technically be a `POST` but I could be wrong

**Comment by tlrobinson** [comment]
I've implemented the deactivated tab and reactivate action. Still need a way to see when the user was deactivated (include `updated_at`?) as well as surface errors when editing a user.

**Comment by tlrobinson** [comment]
@senior Another problem is some errors in `PUT /api/user/:id` such as the `Email address already in use.` error don't follow our standard form error convention like `{ errors: { email: "Email address already in use." }}`

**Comment by tlrobinson** [comment]
Also, `POST /api/user` doesn't return the `is_active` field so new users are appearing in the deactivated tab.

File: src/metabase/api/user.clj [hunk 2]
```diff
@@ -80,7 +73,8 @@
   ;; only allow updates if the specified account is active
   (api/check-404 (db/exists? User, :id id, :is_active true))
   ;; can't change email if it's already taken BY ANOTHER ACCOUNT
-  (api/check-400 (not (db/exists? User, :email email, :id [:not= id])))
+  (api/check (not (db/exists? User, :email email, :id [:not= id]))
+    [400 "Email address already associated to another user."])
```
Comment (tlrobinson) [inline src/metabase/api/user.clj]: This should conform to our errors conventions, either `{ errors: { email: "Email address ..." } }` if it's relevant to a specific field or `{ message: "Email address ..." }` if it applies to the whole request (the former in this case)

**Review by tlrobinson — CHANGES_REQUESTED** [review]

File: src/metabase/api/user.clj [hunk 3]
```diff
@@ -20,42 +20,35 @@
...
-      :is_active     true
-      :is_superuser  false
-      ;; if the user orignally logged in via Google Auth and it's no longer enabled, convert them into a regular user
-      ;; (see Issue #3323)
-      :google_auth   (boolean (and (:google_auth existing-user)
-                                   ;; if google-auth-client-id is set it means Google Auth is enabled
-                                   (session-api/google-auth-client-id)))
-      :ldap_auth     (boolean (and (:ldap_auth existing-user)
-                                   (ldap/ldap-configured?)))))
+  (db/select [User :id :first_name :last_name :email :is_superuser :google_auth :ldap_auth :last_login :is_active]))
```
Comment (tlrobinson) [inline src/metabase/api/user.clj]: How can we ensure `is_active` is returned from the `POST /api/user` and `PUT /api/user/:id` endpoints as well?

File: src/metabase/api/user.clj (hunk 2 above)
Comment (tlrobinson) [reply src/metabase/api/user.clj]: Copied from conversation with @camsaul:

> Writing all the stuff out is annoying so stick a helper function in `metabase.api.common` util like

```
(defn check-field [condition field-kw error-message]
  (when condition
    (throw (ex-info error-message {:status-code 400, :errors {field-kw error-message}}))))
```

> then you could do

```
(check-field (not (db/exists? User :email email)) :email (tru "Email address already in use."))
```

**Comment by senior** [comment]
@tlrobinson I added an `updated_at` field to `core_user`. `is_active` should now be included in the POST and PUT responses and the error messages should be fixed up as you suggested.

File: src/metabase/api/user.clj [hunk 4]
```diff
@@ -19,43 +19,42 @@
...
    last_name  su/NonBlankString
    email      su/Email}
   (api/check-superuser)
-  (if-let [existing-user (db/select-one [User :id :is_active :google_auth], :email email)]
-    ;; this user already exists but is inactive, so simply reactivate the account
-    (reactivate-user! existing-user first_name last_name)
-    ;; new user account, so create it
-    (user/invite-user! first_name last_name email password @api/*current-user*)))
-
+  (api/checkp (not (db/exists? User :email email))
```
Comment (senior) [inline src/metabase/api/user.clj]: @camsaul I wasn't aware of https://github.com/metabase/metabase/pull/7481#discussion_r186896077. Am I using the functions correctly here to get our standard error messages?

File: src/metabase/api/user.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/api/user.clj]: omg @tlrobinson I'm a moron I forgot we already have a function that does exactly that. It's called `api/checkp`. It takes the exact same params

File: src/metabase/api/user.clj (hunk 4 above)
Comment (camsaul) [reply src/metabase/api/user.clj]: Yeah that looks right. I totally forgot this function existed.

Comment (camsaul) [reply src/metabase/api/user.clj]: It was previously more of an informal convention than anything else, however the `defendpoint` param validation code all returns errors in that format so we might as well make it consistent and try to do it everywhere

**Comment by senior** [comment]
Changed `GET /api/user` to return only active users by default. If `include_deactivated` is true, will include all users.

I also merge master in as there were a couple of conflicts. One of the conflicts was the migration I just added for `core_user.updated_at`. If you ran that migration before the merge you'll need to fix it up as it had an id of 76 before and now has an id of 77.

**Comment by tlrobinson** [comment]
I've updated this with the things @senior added.

@senior could you merge `user.clj` again? Thanks.

@kdoh I'm not sure what's going on with this modal scrolling, I'm pretty sure it wasn't like that when I implemented it. Related to any of the recent flex changes?

screenshot 2018-05-17 14 40 46

**Review by tlrobinson — APPROVED** [review]

**Comment by tlrobinson** [comment]
@senior it also looks like the backend tests are failing

**Comment by kdoh** [comment]
@tlrobinson Yeah if I had to guess it's probably another nested `flex-full` that didn't actually need to be there. I can look into it if you'd like.

**Comment by tlrobinson** [comment]
@senior It looks like `is_active` still wasn't being included in the `PUT /api/user`
PUT /api/user` response. I added it in my latest commit by just adding it to `default-user-fields` but that breaks a ton of backend test. If there's a better way could you add it to `PUT /api/user` (and revert my change)?

**Comment by senior** [comment]
@tlrobinson Could your backend be out of date? Just looked through the code and I have a test that covers this and it looks like it should all work. Maybe we can troubleshoot it together and figure out what's missing.

**Comment by tlrobinson** [comment]
Confirmed the last TODO is fixed. Also fixed a small bug with the cancel button not working. Once CI is green this should be ready.

**Pull request merged**